File: src/shared/lens/lens-profile.ts
/**
 * A lens profile as a PICTURE carries it — `RollPicture.lensProfile`.
 *
 * What is stored is the RESOLVED terms (`LensProfileTerms`, the lens pass's
 * own units, for this picture's focal length and aperture), with the names
 * they came from. Not a reference to the database: the export, a second
 * device and a `.roll.json` then draw the same correction with no fetch and no
 * cache, and a later change to Lensfun never moves a picture already
 * developed — preview = export by construction.
 *
 * **A profile is CALIBRATION, not an edit** — the rule the RAW base follows
 * (`picture-sections.ts`: "the RAW base is material, not an edit"). It is a
 * fact about the glass that took the picture, so it is not copied to another
 * picture by the settings sheet (another lens, another focal length), not
 * cleared by Reset, and not what makes a picture "edited" for the delivery
 * rule; it IS part of the export fingerprint, since it changes the file.
 *
 * **Where it applies.** Lensfun measures the lens on RAW data. A camera's own
 * JPEG — and the render inside a RAW — is often corrected in the body already
 * (Sony's *Distortion Comp.: Auto* is on by default), and correcting it again
 * would bend it the other way. So a profile applies on the SENSOR by itself,
 * and on a render only where the author said so (`onRender`), knowing that.
 *
 * Pure.
 */

import { isIdentityProfile, type LensProfileTerms } from '../render/lens';

export interface LensProfileApplied {
  /** `Sony FE 24-70mm f/4 ZA OSS` — the database's lens, maker and model. */
  lens: string;
  /** `Sony ILCE-7CM2` — the body it was matched through. */
  camera: string;
  /** The picture's own, which the terms were interpolated at. */
  focal: number;
  aperture: number | null;
  terms: LensProfileTerms;
  /** What the database had for this lens: a part it lacked is the identity in `terms`. */
  has: { distortion: boolean; tca: boolean; vignette: boolean };
  /** Applied on a camera RENDER too, by the author's choice — see above. */
  onRender: boolean;
}

function finite(v: unknown): number | null {
  return typeof v === 'number' && Number.isFinite(v) ? v : null;
}

function tuple(v: unknown, n: number, fallback: readonly number[]): number[] | null {
  if (!Array.isArray(v) || v.length !== n) return null;
  const out = v.map((x, i) => finite(x) ?? fallback[i]);
  return out;
}

/**
 * A stored profile read back — `undefined` when the field is absent (never
 * decided: an automatic lookup may still apply one), `null` when the author
 * took it off (the lookup never puts it back), else the profile, or null for
 * a record too broken to draw.
 */
export function readLensProfile(raw: unknown): LensProfileApplied | null | undefined {
  if (raw === undefined) return undefined;
  if (!raw || typeof raw !== 'object') return null;
  const src = raw as Record<string, unknown>;
  const t = (src.terms ?? {}) as Record<string, unknown>;
  const distortion = tuple(t.distortion, 4, [0, 0, 0, 0]);
  const tcaRed = tuple(t.tcaRed, 3, [1, 0, 0]);
  const tcaBlue = tuple(t.tcaBlue, 3, [1, 0, 0]);
  const vignette = tuple(t.vignette, 3, [0, 0, 0]);
  const focal = finite(src.focal);
  if (!distortion || !tcaRed || !tcaBlue || !vignette || focal === null || typeof src.lens !== 'string') return null;
  const has = (src.has ?? {}) as Record<string, unknown>;
  return {
    lens: src.lens.slice(0, 120),
    camera: typeof src.camera === 'string' ? src.camera.slice(0, 120) : '',
    focal,
    aperture: finite(src.aperture),
    terms: {
      distortion: distortion as LensProfileTerms['distortion'],
      tcaRed: tcaRed as LensProfileTerms['tcaRed'],
      tcaBlue: tcaBlue as LensProfileTerms['tcaBlue'],
      vignette: vignette as LensProfileTerms['vignette'],
    },
    has: { distortion: has.distortion === true, tca: has.tca === true, vignette: has.vignette === true },
    onRender: src.onRender === true,
  };
}

/** The terms a picture's render draws with: its profile where it applies, else none. */
export function profileInEffect(
  profile: LensProfileApplied | null | undefined,
  onSensor: boolean,
): LensProfileTerms | null {
  if (!profile || isIdentityProfile(profile.terms)) return null;
  return onSensor || profile.onRender ? profile.terms : null;
}

/** `distortion · TCA · vignetting` — what a profile corrects, as a line says it. */
export function describeProfileParts(has: LensProfileApplied['has']): string {
  const parts = [has.distortion && 'distortion', has.tca && 'fringing', has.vignette && 'vignetting'].filter(Boolean);
  return parts.length ? parts.join(' · ') : 'nothing measured';
}

/** The key a lookup is cached under: the body and the lens, as the EXIF names them. */
export function lensKey(make: string, model: string, lensModel: string | null | undefined): string {
  const squash = (s: string) => s.toLowerCase().replace(/[^a-z0-9.]/g, '');
  return `${squash(make)}|${squash(model)}|${squash(lensModel ?? '')}`;
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Lens/LensProfile.swift
// A lens profile as a PICTURE carries it — `RollPicture.lensProfile`. Port of
// `src/shared/lens/lens-profile.ts`.
//
// What is stored is the RESOLVED terms (`LensProfileTerms`, the lens pass's own
// units, for this picture's focal length and aperture), with the names they
// came from. Not a reference to the database: the export, a second device and
// a `.roll.json` then draw the same correction with no fetch and no cache, and
// a later change to Lensfun never moves a picture already developed — preview
// = export by construction.
//
// **A profile is CALIBRATION, not an edit** — the rule the RAW base follows.
// It is a fact about the glass, so it is not copied to another picture by the
// settings sheet, not cleared by Reset, and not what makes a picture "edited"
// for the delivery rule; it IS part of the export fingerprint.
//
// **Where it applies.** Lensfun measures the lens on RAW data. A camera's own
// JPEG — and the render inside a RAW — is often corrected in the body already,
// and correcting it again would bend it the other way. So a profile applies on
// the SENSOR by itself, and on a render only where the author said so
// (`onRender`).
//
// Three states, kept apart exactly as the web keeps them: ABSENT (never
// decided: an automatic lookup may still apply one), NULL (the author took it
// off — or a record too broken to draw), and a profile. `readLensProfile`
// answers with a double optional: `.none` absent, `.some(nil)` null.

import Foundation

public struct LensProfileApplied: Equatable, Sendable {
    /// `Sony FE 24-70mm f/4 ZA OSS` — the database's lens, maker and model.
    public var lens: String
    /// `Sony ILCE-7CM2` — the body it was matched through.
    public var camera: String
    /// The picture's own, which the terms were interpolated at.
    public var focal: Double
    public var aperture: Double?
    public var terms: LensProfileTerms
    /// What the database had for this lens: a part it lacked is the identity in `terms`.
    public var has: LensProfileHas
    /// Applied on a camera RENDER too, by the author's choice — see above.
    public var onRender: Bool

    public init(lens: String, camera: String, focal: Double, aperture: Double?, terms: LensProfileTerms,
                has: LensProfileHas, onRender: Bool) {
        self.lens = lens; self.camera = camera; self.focal = focal; self.aperture = aperture
        self.terms = terms; self.has = has; self.onRender = onRender
    }

    /// The record as the document holds it — what `JSON.stringify` writes of
    /// the web's object (a missing aperture is `null`).
    public var json: JSONValue {
        func list(_ v: [Double]) -> JSONValue { .array(v.map { .number($0) }) }
        return .object([
            "lens": .string(lens),
            "camera": .string(camera),
            "focal": .number(focal),
            "aperture": aperture.map { .number($0) } ?? .null,
            "terms": .object([
                "distortion": list(terms.distortion),
                "tcaRed": list(terms.tcaRed),
                "tcaBlue": list(terms.tcaBlue),
                "vignette": list(terms.vignette),
            ]),
            "has": .object([
                "distortion": .bool(has.distortion),
                "tca": .bool(has.tca),
                "vignette": .bool(has.vignette),
            ]),
            "onRender": .bool(onRender),
        ])
    }
}

private func finite(_ v: JSONValue?) -> Double? {
    v?.finiteNumber
}

private func tuple(_ v: JSONValue?, _ n: Int, _ fallback: [Double]) -> [Double]? {
    guard let a = v?.arrayValue, a.count == n else { return nil }
    return a.enumerated().map { i, x in finite(x) ?? fallback[i] }
}

/// The first `n` UTF-16 code units, as `String.prototype.slice(0, n)`.
private func slice(_ s: String, _ n: Int) -> String {
    if s.utf16.count <= n { return s }
    return String(decoding: Array(s.utf16.prefix(n)), as: UTF16.self)
}

/// A stored profile read back — `.none` when the field is absent (never
/// decided), `.some(nil)` when the author took it off or the record is too
/// broken to draw, else the profile.
public func readLensProfile(_ raw: JSONValue?) -> LensProfileApplied?? {
    guard let raw else { return .none }
    guard let src = raw.objectValue else { return .some(nil) }
    let t = src["terms"]?.objectValue ?? [:]
    let distortion = tuple(t["distortion"], 4, [0, 0, 0, 0])
    let tcaRed = tuple(t["tcaRed"], 3, [1, 0, 0])
    let tcaBlue = tuple(t["tcaBlue"], 3, [1, 0, 0])
    let vignette = tuple(t["vignette"], 3, [0, 0, 0])
    let focal = finite(src["focal"])
    guard let distortion, let tcaRed, let tcaBlue, let vignette, let focal,
          let lens = src["lens"]?.stringValue else { return .some(nil) }
    let has = src["has"]?.objectValue ?? [:]
    return .some(LensProfileApplied(
        lens: slice(lens, 120),
        camera: src["camera"]?.stringValue.map { slice($0, 120) } ?? "",
        focal: focal,
        aperture: finite(src["aperture"]),
        terms: LensProfileTerms(distortion: distortion, tcaRed: tcaRed, tcaBlue: tcaBlue, vignette: vignette),
        has: LensProfileHas(
            distortion: has["distortion"] == .bool(true),
            tca: has["tca"] == .bool(true),
            vignette: has["vignette"] == .bool(true)
        ),
        onRender: src["onRender"] == .bool(true)
    ))
}

/// The terms a picture's render draws with: its profile where it applies, else none.
public func profileInEffect(_ profile: LensProfileApplied?, _ onSensor: Bool) -> LensProfileTerms? {
    guard let profile, !isIdentityProfile(profile.terms) else { return nil }
    return onSensor || profile.onRender ? profile.terms : nil
}

/// `distortion · fringing · vignetting` — what a profile corrects, as a line says it.
public func describeProfileParts(_ has: LensProfileHas) -> String {
    var parts: [String] = []
    if has.distortion { parts.append("distortion") }
    if has.tca { parts.append("fringing") }
    if has.vignette { parts.append("vignetting") }
    return parts.isEmpty ? "nothing measured" : parts.joined(separator: " · ")
}

/// The key a lookup is cached under: the body and the lens, as the EXIF names them.
public func lensKey(_ make: String, _ model: String, _ lensModel: String?) -> String {
    func squash(_ s: String) -> String {
        var out = String.UnicodeScalarView()
        for c in s.lowercased().unicodeScalars
        where (c.value >= 0x61 && c.value <= 0x7A) || (c.value >= 0x30 && c.value <= 0x39) || c == "." {
            out.append(c)
        }
        return String(out)
    }
    return "\(squash(make))|\(squash(model))|\(squash(lensModel ?? ""))"
}
